In [1]:
import numpy as np
import pandas as pd
from scipy import stats
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import datetime
import calendar

In [2]:
# Выгрузим датасет с данными клиентов
customers = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-k-bogatyrev-38/Lessons/olist_customers_dataset.csv')

In [3]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [4]:
# Проверим на наличие неизвестных значений
customers.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [5]:
# Пропусков не обнаружено

In [6]:
# Проверим количество уникальных пользователей
customers.customer_unique_id.nunique()

96096

In [7]:
# Проверим типы данных стобцов
customers.dtypes

customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

In [8]:
# Выгрузим данные по заказам
orders = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-k-bogatyrev-38/Lessons/olist_orders_dataset.csv')

In [9]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [10]:
# Проверим типы данных
orders.dtypes

order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

In [11]:
# Преобразуем столбцы с датами к типу данных date
orders[['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 
        'order_delivered_customer_date', 'order_estimated_delivery_date']] = orders[['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 
        'order_delivered_customer_date', 'order_estimated_delivery_date']].apply(pd.to_datetime)

In [12]:
orders.dtypes

order_id                                 object
customer_id                              object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object

In [13]:
# Проверим пропуски в столбцах
orders.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [14]:
# Выгрузим данные по товарам
items = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-k-bogatyrev-38/Lessons/olist_order_items_dataset.csv')

In [15]:
items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [16]:
# Проверим типы данных
items.dtypes

order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

In [17]:
# Преобразуем столбцы с датами к типу данных date
items['shipping_limit_date'] = items['shipping_limit_date'].apply(pd.to_datetime)

In [18]:
items.dtypes

order_id                       object
order_item_id                   int64
product_id                     object
seller_id                      object
shipping_limit_date    datetime64[ns]
price                         float64
freight_value                 float64
dtype: object

In [19]:
# Проверим наличие пропусков в столбцах
items.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

# Task 1: Сколько у нас пользователей, которые совершили покупку только один раз?

## Заказ будет считаться купленным, если он доставлен и есть подтвержденное время оплаты

## Отберем заказы, удовлетворяющие нашим требованиям покупки

In [20]:
delivered = orders.query("order_status == 'delivered'")

In [21]:
delivered = delivered.dropna(subset=['order_approved_at'])

In [22]:
delivered.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     2
order_delivered_customer_date    8
order_estimated_delivery_date    0
dtype: int64

In [23]:
# Совместим датасеты по клиентам и купленным заказам
customers_with_1_purchase = customers.merge(delivered, on='customer_id')

In [24]:
customers_with_1_purchase.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date  
0           2017-05-25 10:35:35                    2017-06-05  
1           2018-01-29 12:41:19                    2018-02-06  
2           2018-06-14 17:58:51                    2018-06-13  
3           2018-03-28 16:04:25                    2018-04-10  
4           2018-08-09 20:55:48                    2018-08-15

In [25]:
# Сгруппируем данные по уникальному идентификатору клиента и количеству покупок
customers_with_1_purchase = customers_with_1_purchase.groupby('customer_unique_id', as_index=False)\
                         .agg({'order_id':'count'})

In [26]:
# Переименуем колонку
customers_with_1_purchase.rename(columns={'order_id':'purchase_quantity'}, inplace=True)

In [27]:
# Отфильтруем полученные значения со значением количества покупок = 1
customers_with_1_purchase.query('purchase_quantity == 1', inplace=True)

In [28]:
# Выведем количество пользователей, совершивших 1 покупку
customers_with_1_purchase.customer_unique_id.count()

90545

## Ответ 90545

# Task 2: Сколько заказов в месяц в среднем не доставляется по разным причинам (вывести детализацию по причинам)?

## Выведем недоставленные заказы

In [29]:
not_delivered_orders = orders.query("order_status != 'delivered'")

In [30]:
not_delivered_orders.head()

order_id                       customer_id  \
6    136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44   ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103  0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128  15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154  6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   

    order_status order_purchase_timestamp   order_approved_at  \
6       invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
44       shipped      2018-06-04 16:44:48 2018-06-05 04:31:18   
103     invoiced      2018-08-03 17:44:42 2018-08-07 06:15:14   
128   processing      2017-09-03 14:22:03 2017-09-03 14:30:09   
154      shipped      2018-01-10 11:33:07 2018-01-11 02:32:30   

    order_delivered_carrier_date order_delivered_customer_date  \
6                            NaT                           NaT   
44           2018-06-05 14:32:00                           NaT   
103                          NaT                           NaT   
128                          NaT                           NaT   
154          2018-01-11 19:39:23                           NaT   

    order_estimated_delivery_date  
6                      2017-05-09  
44                     2018-06-28  
103                    2018-08-21  
128                    2017-10-03  
154                    2018-02-07

In [31]:
# Добавим новый столбец с месяцом в котором был сделан заказ
not_delivered_orders['order_month'] = not_delivered_orders.order_purchase_timestamp.dt.month

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [32]:
# Добавим новый столбец с годом в котором был сделан заказ
not_delivered_orders['order_year'] = not_delivered_orders.order_purchase_timestamp.dt.year

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Недоставленными заказами будем считать со статусами canceled и unavailable, так как остальные могут находиться в процессе формирования или в работе

In [33]:
# Преобразуем наш дата фрейм к этим заказам
not_delivered_orders = not_delivered_orders.query("order_status in ('canceled', 'unavailable')")

In [34]:
# Сгруппируем данные по заказам со статусом canceled по году и месяцу, в который был сделан заказ и высчитаем среднее
canceled = not_delivered_orders.query("order_status == 'canceled'")\
                               .groupby(['order_year', 'order_month'], as_index=False)\
                               .agg({'order_id':'count'})

In [35]:
canceled_per_month = canceled.order_id.mean().round(2)

In [36]:
canceled_per_month

26.04

In [37]:
# Проделаем то же самое с заказами со статусом unavailable
unavailable = not_delivered_orders.query("order_status == 'unavailable'")\
                               .groupby(['order_year', 'order_month'], as_index=False)\
                               .agg({'order_id':'count'})

In [38]:
unavailable_per_month = unavailable.order_id.mean().round(2)

In [39]:
unavailable_per_month

29.0

## Ответ: unavailable=29.0; canceled=26.04

# Task 3: По каждому товару определить, в какой день недели товар чаще всего покупается.

In [40]:
delivered.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

In [41]:
items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value  
0 2017-09-19 09:45:35   58.90          13.29  
1 2017-05-03 11:05:13  239.90          19.93  
2 2018-01-18 14:48:30  199.00          17.87  
3 2018-08-15 10:10:18   12.99          12.79  
4 2017-02-13 13:57:51  199.90          18.14

In [42]:
# Создадим новый датасет, в который включены информация о товарах и заказах

In [43]:
delivered_items = delivered.merge(items, on='order_id')

In [44]:
delivered_items.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0                    2017-10-18              1   
1                    2018-08-13              1   
2                    2018-09-04              1   
3                    2017-12-15              1   
4                    2018-02-26              1   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
2  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   
3  d0b61bfb1de832b15ba9d266ca96e5b0  66922902710d126a0e7d26b0e3805106   
4  65266b2da20d04dbe00c5c2d3bb7859e  2c9e548be18521d1c43cde1c582c6de8   

  shipping_limit_date   price  freight_value  
0 2017-10-06 11:07:15   29.99           8.72  
1 2018-07-30 03:24:27  118.70          22.76  
2 2018-08-13 08:55:23  159.90          19.22  
3 2017-11-23 19:45:59   45.00          27.20  
4 2018-02-19 20:31:37   19.90           8.72

In [45]:
# Проверим типы данных
delivered_items.dtypes

order_id                                 object
customer_id                              object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
order_item_id                             int64
product_id                               object
seller_id                                object
shipping_limit_date              datetime64[ns]
price                                   float64
freight_value                           float64
dtype: object

In [46]:
# Создадим новый столбец с днем недели покупки
delivered_items['weekday_purchase'] = delivered_items.order_purchase_timestamp.dt.day_name()

In [47]:
delivered_items.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0                    2017-10-18              1   
1                    2018-08-13              1   
2                    2018-09-04              1   
3                    2017-12-15              1   
4                    2018-02-26              1   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
2  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   
3  d0b61bfb1de832b15ba9d266ca96e5b0  66922902710d126a0e7d26b0e3805106   
4  65266b2da20d04dbe00c5c2d3bb7859e  2c9e548be18521d1c43cde1c582c6de8   

  shipping_limit_date   price  freight_value weekday_purchase  
0 2017-10-06 11:07:15   29.99           8.72           Monday  
1 2018-07-30 03:24:27  118.70          22.76          Tuesday  
2 2018-08-13 08:55:23  159.90          19.22        Wednesday  
3 2017-11-23 19:45:59   45.00          27.20         Saturday  
4 2018-02-19 20:31:37   19.90           8.72          Tuesday

In [48]:
# Сгруппируем данные по идентификатору продукта и дню недели его покупки
delivered_items = delivered_items.groupby(['product_id', 'weekday_purchase'], as_index=False).agg({'order_id':'count'})

In [49]:
delivered_items.head()

product_id weekday_purchase  order_id
0  00066f42aeeb9f3007548bb9d3f33c38           Sunday         1
1  00088930e925c41fd95ebfe695fd2655          Tuesday         1
2  0009406fd7479715e4bef61dd91f2462         Thursday         1
3  000b8f95fcb9e0096488278317764d19           Friday         1
4  000b8f95fcb9e0096488278317764d19        Wednesday         1

In [50]:
# Переименуем колонки
delivered_items.rename(columns={'order_id':'quantity'}, inplace=True)

In [51]:
delivered_items.head()

product_id weekday_purchase  quantity
0  00066f42aeeb9f3007548bb9d3f33c38           Sunday         1
1  00088930e925c41fd95ebfe695fd2655          Tuesday         1
2  0009406fd7479715e4bef61dd91f2462         Thursday         1
3  000b8f95fcb9e0096488278317764d19           Friday         1
4  000b8f95fcb9e0096488278317764d19        Wednesday         1

In [52]:
# Преобразуем таблицу в более удобный вид
delivered_items = delivered_items.pivot(columns='weekday_purchase', index='product_id', values='quantity')

In [53]:
delivered_items

weekday_purchase                  Friday  Monday  Saturday  Sunday  Thursday  \
product_id                                                                     
00066f42aeeb9f3007548bb9d3f33c38     NaN     NaN       NaN     1.0       NaN   
00088930e925c41fd95ebfe695fd2655     NaN     NaN       NaN     NaN       NaN   
0009406fd7479715e4bef61dd91f2462     NaN     NaN       NaN     NaN       1.0   
000b8f95fcb9e0096488278317764d19     1.0     NaN       NaN     NaN       NaN   
000d9be29b5207b54e86aa1b1ac54872     NaN     NaN       NaN     NaN       NaN   
...                                  ...     ...       ...     ...       ...   
fff6177642830a9a94a0f2cba5e476d1     NaN     NaN       1.0     1.0       NaN   
fff81cc3158d2725c0655ab9ba0f712c     NaN     1.0       NaN     NaN       NaN   
fff9553ac224cec9d15d49f5a263411f     1.0     NaN       NaN     NaN       NaN   
fffdb2d0ec8d6a61f0a0a0db3f25b441     1.0     NaN       NaN     1.0       1.0   
fffe9eeff12fcbd74a2f2b007dde0c58     NaN     NaN       NaN     NaN       NaN   

weekday_purchase                  Tuesday  Wednesday  
product_id                                            
00066f42aeeb9f3007548bb9d3f33c38      NaN        NaN  
00088930e925c41fd95ebfe695fd2655      1.0        NaN  
0009406fd7479715e4bef61dd91f2462      NaN        NaN  
000b8f95fcb9e0096488278317764d19      NaN        1.0  
000d9be29b5207b54e86aa1b1ac54872      1.0        NaN  
...                                   ...        ...  
fff6177642830a9a94a0f2cba5e476d1      NaN        NaN  
fff81cc3158d2725c0655ab9ba0f712c      NaN        NaN  
fff9553ac224cec9d15d49f5a263411f      NaN        NaN  
fffdb2d0ec8d6a61f0a0a0db3f25b441      2.0        NaN  
fffe9eeff12fcbd74a2f2b007dde0c58      NaN        1.0  

[32213 rows x 7 columns]

In [54]:
# Найдем день недели в который было совершено наибольшее количество покупок продукта
delivered_items = delivered_items.idxmax(axis=1).reset_index()

In [55]:
# Ответ
delivered_items.rename(columns={0:'weekday_max_purchase'}, inplace=True)

In [56]:
delivered_items

product_id weekday_max_purchase
0      00066f42aeeb9f3007548bb9d3f33c38               Sunday
1      00088930e925c41fd95ebfe695fd2655              Tuesday
2      0009406fd7479715e4bef61dd91f2462             Thursday
3      000b8f95fcb9e0096488278317764d19               Friday
4      000d9be29b5207b54e86aa1b1ac54872              Tuesday
...                                 ...                  ...
32208  fff6177642830a9a94a0f2cba5e476d1             Saturday
32209  fff81cc3158d2725c0655ab9ba0f712c               Monday
32210  fff9553ac224cec9d15d49f5a263411f               Friday
32211  fffdb2d0ec8d6a61f0a0a0db3f25b441              Tuesday
32212  fffe9eeff12fcbd74a2f2b007dde0c58            Wednesday

[32213 rows x 2 columns]

# Task 4: Сколько у каждого из пользователей в среднем покупок в неделю (по месяцам)? Не стоит забывать, что внутри месяца может быть не целое количество недель. Например, в ноябре 2021 года 4,28 недели. И внутри метрики это нужно учесть.

In [57]:
# Сформируем датафрейм
purchase_per_week = customers.merge(delivered, on='customer_id')

In [58]:
# Для точности данных выберем только неотмененные заказы
purchase_per_week = purchase_per_week.query("order_status != 'canceled'")

In [59]:
# Введем новую колонку с периодом покупки
purchase_per_week['month_of_the_year'] = purchase_per_week.order_purchase_timestamp.dt.to_period("M")

In [60]:
# Создадим новый df с группировкой по пользователю и периоду покупки
purchase_per_week = purchase_per_week.groupby(['customer_unique_id', 'month_of_the_year'], as_index=False)\
                                     .agg({'order_id':'count'})

In [61]:
purchase_per_week

customer_unique_id month_of_the_year  order_id
0      0000366f3b9a7992bf8c76cfdf3221e2           2018-05         1
1      0000b849f77a49e4a4ce2b2a4ca5be3f           2018-05         1
2      0000f46a3911fa3c0805444483337064           2017-03         1
3      0000f6ccb0745a6a4b88665a16c9f078           2017-10         1
4      0004aac84e0df4da2b147fca70cf8255           2017-11         1
...                                 ...               ...       ...
95175  fffcf5a5ff07b0908bd4e2dbc735a684           2017-06         1
95176  fffea47cd6d3cc0a88bd621562a9d061           2017-12         1
95177  ffff371b4d645b6ecea244b27531430a           2017-02         1
95178  ffff5962728ec6157033ef9805bacc48           2018-05         1
95179  ffffd2657e2aad2907e67c3e9daecbeb           2017-05         1

[95180 rows x 3 columns]

In [62]:
# Введем новую колонку с количеством недель в месяце
purchase_per_week['weeks_in_month'] = purchase_per_week.month_of_the_year.dt.days_in_month / 7

In [63]:
# Введем новую колонку с количесвом заказов в неделю
purchase_per_week['purchases_in_week'] = purchase_per_week.order_id / purchase_per_week.weeks_in_month

In [64]:
purchase_per_week

customer_unique_id month_of_the_year  order_id  \
0      0000366f3b9a7992bf8c76cfdf3221e2           2018-05         1   
1      0000b849f77a49e4a4ce2b2a4ca5be3f           2018-05         1   
2      0000f46a3911fa3c0805444483337064           2017-03         1   
3      0000f6ccb0745a6a4b88665a16c9f078           2017-10         1   
4      0004aac84e0df4da2b147fca70cf8255           2017-11         1   
...                                 ...               ...       ...   
95175  fffcf5a5ff07b0908bd4e2dbc735a684           2017-06         1   
95176  fffea47cd6d3cc0a88bd621562a9d061           2017-12         1   
95177  ffff371b4d645b6ecea244b27531430a           2017-02         1   
95178  ffff5962728ec6157033ef9805bacc48           2018-05         1   
95179  ffffd2657e2aad2907e67c3e9daecbeb           2017-05         1   

       weeks_in_month  purchases_in_week  
0            4.428571           0.225806  
1            4.428571           0.225806  
2            4.428571           0.225806  
3            4.428571           0.225806  
4            4.285714           0.233333  
...               ...                ...  
95175        4.285714           0.233333  
95176        4.428571           0.225806  
95177        4.000000           0.250000  
95178        4.428571           0.225806  
95179        4.428571           0.225806  

[95180 rows x 5 columns]

In [65]:
# Вычислим среднее по каждому пользователю
purchase_per_week = purchase_per_week.groupby('customer_unique_id', as_index=False).agg({'purchases_in_week':'mean'})

# Ответ

In [66]:
purchase_per_week

customer_unique_id  purchases_in_week
0      0000366f3b9a7992bf8c76cfdf3221e2           0.225806
1      0000b849f77a49e4a4ce2b2a4ca5be3f           0.225806
2      0000f46a3911fa3c0805444483337064           0.225806
3      0000f6ccb0745a6a4b88665a16c9f078           0.225806
4      0004aac84e0df4da2b147fca70cf8255           0.233333
...                                 ...                ...
93340  fffcf5a5ff07b0908bd4e2dbc735a684           0.233333
93341  fffea47cd6d3cc0a88bd621562a9d061           0.225806
93342  ffff371b4d645b6ecea244b27531430a           0.250000
93343  ffff5962728ec6157033ef9805bacc48           0.225806
93344  ffffd2657e2aad2907e67c3e9daecbeb           0.225806

[93345 rows x 2 columns]

# Task 5: Используя pandas, проведи когортный анализ пользователей. В период с января по декабрь выяви когорту с самым высоким retention на 3й месяц.

In [67]:
# Сформируем общий дата фрейм
df_prom = customers.merge(delivered, on= 'customer_id')

In [68]:
df = df_prom.merge(items, on='order_id')

In [69]:
df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date  order_item_id  \
0           2017-05-25 10:35:35                    2017-06-05              1   
1           2018-01-29 12:41:19                    2018-02-06              1   
2           2018-06-14 17:58:51                    2018-06-13              1   
3           2018-03-28 16:04:25                    2018-04-10              1   
4           2018-08-09 20:55:48                    2018-08-15              1   

                         product_id                         seller_id  \
0  a9516a079e37a9c9c36b9b78b10169e8  7c67e1448b00f6e969d365cea6b010ab   
1  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
2  bd07b66896d6f1494f5b86251848ced7  7c67e1448b00f6e969d365cea6b010ab   
3  a5647c44af977b148e0a3a4751a09e2e  7c67e1448b00f6e969d365cea6b010ab   
4  9391a573abe00141c56e38d84d7d5b3b  4a3ca9315b744ce9f8e9374361493884   

  shipping_limit_date   price  freight_value  
0 2017-05-22 15:22:12  124.99          21.88  
1 2018-01-18 20:58:32  289.00          46.48  
2 2018-06-05 16:19:10  139.94          17.79  
3 2018-03-27 16:31:16  149.94          23.36  
4 2018-07-31 10:10:09  230.00          22.25

In [70]:
# Сформируем периоды покупок
df['period'] = df.order_approved_at.dt.strftime('%Y-%m')

In [71]:
df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date  order_item_id  \
0           2017-05-25 10:35:35                    2017-06-05              1   
1           2018-01-29 12:41:19                    2018-02-06              1   
2           2018-06-14 17:58:51                    2018-06-13              1   
3           2018-03-28 16:04:25                    2018-04-10              1   
4           2018-08-09 20:55:48                    2018-08-15              1   

                         product_id                         seller_id  \
0  a9516a079e37a9c9c36b9b78b10169e8  7c67e1448b00f6e969d365cea6b010ab   
1  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
2  bd07b66896d6f1494f5b86251848ced7  7c67e1448b00f6e969d365cea6b010ab   
3  a5647c44af977b148e0a3a4751a09e2e  7c67e1448b00f6e969d365cea6b010ab   
4  9391a573abe00141c56e38d84d7d5b3b  4a3ca9315b744ce9f8e9374361493884   

  shipping_limit_date   price  freight_value   period  
0 2017-05-22 15:22:12  124.99          21.88  2017-05  
1 2018-01-18 20:58:32  289.00          46.48  2018-01  
2 2018-06-05 16:19:10  139.94          17.79  2018-05  
3 2018-03-27 16:31:16  149.94          23.36  2018-03  
4 2018-07-31 10:10:09  230.00          22.25  2018-07

In [72]:
# Сформируем период первой покупки
df.set_index('customer_unique_id', inplace=True)

In [73]:
# Отсортируем данные с января по декабрь
df = df.query("period >= '2017-01' and period < '2018-01'")

In [74]:
df['first_purchase_period'] = df.groupby(level=0)['order_approved_at'].min().dt.strftime('%Y-%m')

In [75]:
# Создадим когорты
df.reset_index(inplace=True)

In [76]:
cohorts = df.groupby(['first_purchase_period', 'period']).agg({'customer_unique_id':'nunique', 'order_id':'count', 'price':'sum'})

In [77]:
cohorts.head()

customer_unique_id  order_id      price
first_purchase_period period                                          
2017-01               2017-01                 684       869  106888.10
                      2017-02                   3         3     125.12
                      2017-03                   2         2      82.89
                      2017-04                   1         1      75.00
                      2017-05                   3         3     199.70

In [78]:
# Переименуем колонки
cohorts.rename(columns={'customer_unique_id':'total_customers', 'order_id':'total_orders', 'price':'total_price'}, inplace=True)

In [79]:
cohorts

total_customers  total_orders  total_price
first_purchase_period period                                             
2017-01               2017-01              684           869    106888.10
                      2017-02                3             3       125.12
                      2017-03                2             2        82.89
                      2017-04                1             1        75.00
                      2017-05                3             3       199.70
...                                        ...           ...          ...
2017-10               2017-11               30            38      3376.74
                      2017-12               13            16      1393.71
2017-11               2017-11             6928          8176    958317.08
                      2017-12               39            52      6195.37
2017-12               2017-12             5490          6229    731518.12

[77 rows x 3 columns]

In [80]:
# Введем формулу для упорядочивания когортов
def cohort_period(df_exp):
    df_exp['cohort_period'] = np.arange(len(df_exp)) + 1
    return df_exp

In [81]:
# Применим данную формулу для наших когортов
cohorts = cohorts.groupby(level=0).apply(cohort_period)

In [82]:
cohorts

total_customers  total_orders  total_price  \
first_purchase_period period                                                
2017-01               2017-01              684           869    106888.10   
                      2017-02                3             3       125.12   
                      2017-03                2             2        82.89   
                      2017-04                1             1        75.00   
                      2017-05                3             3       199.70   
...                                        ...           ...          ...   
2017-10               2017-11               30            38      3376.74   
                      2017-12               13            16      1393.71   
2017-11               2017-11             6928          8176    958317.08   
                      2017-12               39            52      6195.37   
2017-12               2017-12             5490          6229    731518.12   

                               cohort_period  
first_purchase_period period                  
2017-01               2017-01              1  
                      2017-02              2  
                      2017-03              3  
                      2017-04              4  
                      2017-05              5  
...                                      ...  
2017-10               2017-11              2  
                      2017-12              3  
2017-11               2017-11              1  
                      2017-12              2  
2017-12               2017-12              1  

[77 rows x 4 columns]

In [83]:
# Переиндексируем наш df по периоду когорта и периоду первой покупки
cohorts.reset_index(inplace=True)

In [84]:
cohorts.set_index(['cohort_period', 'first_purchase_period'], inplace=True)

In [85]:
cohorts

period  total_customers  total_orders  \
cohort_period first_purchase_period                                           
1             2017-01                2017-01              684           869   
2             2017-01                2017-02                3             3   
3             2017-01                2017-03                2             2   
4             2017-01                2017-04                1             1   
5             2017-01                2017-05                3             3   
...                                      ...              ...           ...   
2             2017-10                2017-11               30            38   
3             2017-10                2017-12               13            16   
1             2017-11                2017-11             6928          8176   
2             2017-11                2017-12               39            52   
1             2017-12                2017-12             5490          6229   

                                     total_price  
cohort_period first_purchase_period               
1             2017-01                  106888.10  
2             2017-01                     125.12  
3             2017-01                      82.89  
4             2017-01                      75.00  
5             2017-01                     199.70  
...                                          ...  
2             2017-10                    3376.74  
3             2017-10                    1393.71  
1             2017-11                  958317.08  
2             2017-11                    6195.37  
1             2017-12                  731518.12  

[77 rows x 4 columns]

In [86]:
# Выделим значения для сравнения для каждого периода первой покупки
cohort_period_size = cohorts.groupby(level=1)['total_customers'].first()

In [87]:
cohort_period_size

first_purchase_period
2017-01     684
2017-02    1612
2017-03    2511
2017-04    2233
2017-05    3453
2017-06    3044
2017-07    3710
2017-08    4081
2017-09    4026
2017-10    4294
2017-11    6928
2017-12    5490
Name: total_customers, dtype: int64

In [88]:
# Преобразуем нашу таблицу, чтобы периоду когортов являлись столбцами, индексами являлись периоды первой покупки, а значениями
# являлось количество клиентов

In [89]:
cohorts['total_customers'].unstack(0)

cohort_period              1     2     3     4     5     6     7    8    9   \
first_purchase_period                                                         
2017-01                 684.0   3.0   2.0   1.0   3.0   1.0   3.0  1.0  1.0   
2017-02                1612.0   3.0   4.0   3.0   7.0   2.0   4.0  3.0  1.0   
2017-03                2511.0  11.0   9.0   9.0   9.0   4.0   4.0  8.0  7.0   
2017-04                2233.0  14.0   5.0   4.0   5.0   6.0   8.0  6.0  8.0   
2017-05                3453.0  18.0  16.0  10.0  10.0  11.0  14.0  5.0  NaN   
2017-06                3044.0  15.0  12.0  13.0   9.0  11.0  11.0  NaN  NaN   
2017-07                3710.0  20.0  12.0   7.0  13.0   7.0   NaN  NaN  NaN   
2017-08                4081.0  28.0  15.0  11.0  15.0   NaN   NaN  NaN  NaN   
2017-09                4026.0  27.0  20.0  13.0   NaN   NaN   NaN  NaN  NaN   
2017-10                4294.0  30.0  13.0   NaN   NaN   NaN   NaN  NaN  NaN   
2017-11                6928.0  39.0   NaN   NaN   NaN   NaN   NaN  NaN  NaN   
2017-12                5490.0   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN   

cohort_period           10   11  
first_purchase_period            
2017-01                2.0  2.0  
2017-02                3.0  2.0  
2017-03                3.0  NaN  
2017-04                NaN  NaN  
2017-05                NaN  NaN  
2017-06                NaN  NaN  
2017-07                NaN  NaN  
2017-08                NaN  NaN  
2017-09                NaN  NaN  
2017-10                NaN  NaN  
2017-11                NaN  NaN  
2017-12                NaN  NaN

In [90]:
# Вычисляем customers retention

In [91]:
customers_retention = cohorts['total_customers'].unstack(0).divide(cohort_period_size, axis=0)

In [92]:
customers_retention

cohort_period           1         2         3         4         5         6   \
first_purchase_period                                                          
2017-01                1.0  0.004386  0.002924  0.001462  0.004386  0.001462   
2017-02                1.0  0.001861  0.002481  0.001861  0.004342  0.001241   
2017-03                1.0  0.004381  0.003584  0.003584  0.003584  0.001593   
2017-04                1.0  0.006270  0.002239  0.001791  0.002239  0.002687   
2017-05                1.0  0.005213  0.004634  0.002896  0.002896  0.003186   
2017-06                1.0  0.004928  0.003942  0.004271  0.002957  0.003614   
2017-07                1.0  0.005391  0.003235  0.001887  0.003504  0.001887   
2017-08                1.0  0.006861  0.003676  0.002695  0.003676       NaN   
2017-09                1.0  0.006706  0.004968  0.003229       NaN       NaN   
2017-10                1.0  0.006986  0.003027       NaN       NaN       NaN   
2017-11                1.0  0.005629       NaN       NaN       NaN       NaN   
2017-12                1.0       NaN       NaN       NaN       NaN       NaN   

cohort_period                7         8         9         10        11  
first_purchase_period                                                    
2017-01                0.004386  0.001462  0.001462  0.002924  0.002924  
2017-02                0.002481  0.001861  0.000620  0.001861  0.001241  
2017-03                0.001593  0.003186  0.002788  0.001195       NaN  
2017-04                0.003583  0.002687  0.003583       NaN       NaN  
2017-05                0.004054  0.001448       NaN       NaN       NaN  
2017-06                0.003614       NaN       NaN       NaN       NaN  
2017-07                     NaN       NaN       NaN       NaN       NaN  
2017-08                     NaN       NaN       NaN       NaN       NaN  
2017-09                     NaN       NaN       NaN       NaN       NaN  
2017-10                     NaN       NaN       NaN       NaN       NaN  
2017-11                     NaN       NaN       NaN       NaN       NaN  
2017-12                     NaN       NaN       NaN       NaN       NaN

# Ответ: максимальный retention на 3-й месяц наблюдается в когорте пользователей 2017-06

In [93]:
customers_retention.loc[customers_retention[4].idxmax()]

cohort_period
1     1.000000
2     0.004928
3     0.003942
4     0.004271
5     0.002957
6     0.003614
7     0.003614
8          NaN
9          NaN
10         NaN
11         NaN
Name: 2017-06, dtype: float64

# Task 6: Выполнить RFM сегментацию пользователей. В кластеризации можешь выбрать следующие метрики: R - время от последней покупки пользователя до текущей даты, F - суммарное количество покупок у пользователя за всё время, M - сумма покупок за всё время.

In [94]:
# Сформируем общий дата фрейм
df_prom = customers.merge(delivered, on= 'customer_id')

In [95]:
df = df_prom.merge(items, on='order_id')

In [96]:
# Определим текущую дату
current_date = df.order_approved_at.max() + pd.to_timedelta(1, 'D')

In [97]:
# Создадим новый df с характеристиками в задании по уникальному идентификатору пользователя
rfm_df = df.groupby('customer_unique_id')\
           .agg({'order_approved_at':lambda x: (current_date - x.max()).days, 'order_id':'count', 'price':'sum'})

In [98]:
# Переименуем колонки
rfm_df.rename(columns={'order_approved_at':'recency', 'order_id':'frequency', 'price':'momentary_revenue'}, inplace=True)

In [99]:
rfm_df

recency  frequency  momentary_revenue
customer_unique_id                                                     
0000366f3b9a7992bf8c76cfdf3221e2      112          1             129.90
0000b849f77a49e4a4ce2b2a4ca5be3f      114          1              18.90
0000f46a3911fa3c0805444483337064      537          1              69.00
0000f6ccb0745a6a4b88665a16c9f078      321          1              25.99
0004aac84e0df4da2b147fca70cf8255      288          1             180.00
...                                   ...        ...                ...
fffcf5a5ff07b0908bd4e2dbc735a684      447          2            1570.00
fffea47cd6d3cc0a88bd621562a9d061      262          1              64.89
ffff371b4d645b6ecea244b27531430a      568          1              89.90
ffff5962728ec6157033ef9805bacc48      119          1             115.00
ffffd2657e2aad2907e67c3e9daecbeb      483          1              56.99

[93345 rows x 3 columns]

In [100]:
# Введем диапозоны распределения искомых величин в виде 4 квантилей
quantiles = rfm_df.quantile([0.2, 0.4, 0.6, 0.8])

In [101]:
quantiles = quantiles.to_dict()

In [102]:
# Применим функцию для определения рейтинга
def score(x,p,d):
     if x <= d[p][0.20]:
         return str(1)
     elif x <= d[p][0.40]:
         return str(2)
     elif x <= d[p][0.60]: 
         return str(3)
     elif x<=d[p][0.80]:
         return str(4)
     else:
         return str(5)

In [103]:
# Создадим новые столбцы с рейтингом величины конкретного пользователя с помощью функции, написанной раннее
rfm_df['r_quantile'] = rfm_df.recency.apply(score, args=('recency', quantiles))

In [104]:
rfm_df['f_quantile'] = rfm_df.frequency.apply(score, args=('frequency', quantiles))

In [105]:
rfm_df['m_quantile'] = rfm_df.momentary_revenue.apply(score, args=('momentary_revenue', quantiles))

In [106]:
rfm_df['score'] = rfm_df.r_quantile + rfm_df.f_quantile + rfm_df.m_quantile

In [107]:
# Введем сегменты
segt_map = {
    r'[1-2][1-2]': 'ненадежный',
    r'[1-2][3-4]': 'в зоне риска потери',
    r'[1-2]5': 'не потеряем (возможно)',
    r'3[1-2]': 'находится с в стадии сна',
    r'33': 'требует внимания',
    r'[3-4][4-5]': 'постоянный клиент',
    r'41': 'обещанный покупатель',
    r'51': 'новый клиент',
    r'[4-5][2-3]': 'потенциально постоянный клиент',
    r'5[4-5]': 'самый постоянный клиент'
}

In [108]:
rfm_df['segment'] = rfm_df['r_quantile'] + rfm_df['f_quantile']

In [109]:
rfm_df['segment'] = rfm_df['segment'].replace(segt_map, regex=True)

# Ответ

In [110]:
rfm_df

recency  frequency  momentary_revenue  \
customer_unique_id                                                        
0000366f3b9a7992bf8c76cfdf3221e2      112          1             129.90   
0000b849f77a49e4a4ce2b2a4ca5be3f      114          1              18.90   
0000f46a3911fa3c0805444483337064      537          1              69.00   
0000f6ccb0745a6a4b88665a16c9f078      321          1              25.99   
0004aac84e0df4da2b147fca70cf8255      288          1             180.00   
...                                   ...        ...                ...   
fffcf5a5ff07b0908bd4e2dbc735a684      447          2            1570.00   
fffea47cd6d3cc0a88bd621562a9d061      262          1              64.89   
ffff371b4d645b6ecea244b27531430a      568          1              89.90   
ffff5962728ec6157033ef9805bacc48      119          1             115.00   
ffffd2657e2aad2907e67c3e9daecbeb      483          1              56.99   

                                 r_quantile f_quantile m_quantile score  \
customer_unique_id                                                        
0000366f3b9a7992bf8c76cfdf3221e2          2          1          4   214   
0000b849f77a49e4a4ce2b2a4ca5be3f          2          1          1   211   
0000f46a3911fa3c0805444483337064          5          1          2   512   
0000f6ccb0745a6a4b88665a16c9f078          4          1          1   411   
0004aac84e0df4da2b147fca70cf8255          4          1          5   415   
...                                     ...        ...        ...   ...   
fffcf5a5ff07b0908bd4e2dbc735a684          5          5          5   555   
fffea47cd6d3cc0a88bd621562a9d061          3          1          2   312   
ffff371b4d645b6ecea244b27531430a          5          1          3   513   
ffff5962728ec6157033ef9805bacc48          2          1          4   214   
ffffd2657e2aad2907e67c3e9daecbeb          5          1          2   512   

                                                   segment  
customer_unique_id                                          
0000366f3b9a7992bf8c76cfdf3221e2                ненадежный  
0000b849f77a49e4a4ce2b2a4ca5be3f                ненадежный  
0000f46a3911fa3c0805444483337064              новый клиент  
0000f6ccb0745a6a4b88665a16c9f078      обещанный покупатель  
0004aac84e0df4da2b147fca70cf8255      обещанный покупатель  
...                                                    ...  
fffcf5a5ff07b0908bd4e2dbc735a684   самый постоянный клиент  
fffea47cd6d3cc0a88bd621562a9d061  находится с в стадии сна  
ffff371b4d645b6ecea244b27531430a              новый клиент  
ffff5962728ec6157033ef9805bacc48                ненадежный  
ffffd2657e2aad2907e67c3e9daecbeb              новый клиент  

[93345 rows x 8 columns]